pdns/dnsdistdist/dnsdist-opentelemetry.hh

```diff
@@ -0,0 +1,318 @@
+using AnyValue = std::variant<std::string, int>;
+using TracesData = int;
+#endif
+
+/*
+ * This namespace contains all the bits and pieces required to do OpenTelemetry
+ * traces in dnsdist. It is contained in this header and cc-file to ensure the rest
+ * of the code is not littered with #ifdefs for DISABLE_PROTOBUF. All functions and
+ * other public members can be safely called/manipulated in a non-protobuf build of
+ * dnsdist.
+ *
+ * The idea is inspired by the rec-eventrace.{cc,hh} files.
+ *
+ * Although the namespace contains dnsdist, it might be general enough to be
+ * reused (after renaming the namespace) by auth and recursor
+ */
+namespace pdns::trace::dnsdist
+{
+
+/**
+ * @class Tracer
+ * @brief This class holds a single trace instance
+ *
+ */
+class Tracer
+{
+public:
+  Tracer()
+  {
+#ifndef DISABLE_PROTOBUF
+    pdns::trace::random(d_traceid);
+#endif
+  }
+  ~Tracer() = default;
+  Tracer(const Tracer&) = delete;
+  Tracer& operator=(const Tracer) = delete;
+  Tracer& operator=(Tracer&&) = delete;
+  Tracer(Tracer&&) = delete;
+
+  /**
+   * @brief Activate the Tracer
+   *
+   * Once activated, all new events are stored as actual Spans
+   */
+  void activate()
+  {
+#ifndef DISABLE_PROTOBUF
+    d_activated = true;
+    if (d_postActivationSpans == nullptr) {
+      d_postActivationSpans = std::make_unique<std::vector<pdns::trace::Span>>();
+    }
+#endif
+  }
+
+  /**
+   * @brief Deactivate the tracer
+   */
+  void deactivate()
+  {
+#ifndef DISABLE_PROTOBUF
+    // Set deactivated, but don't delete any data.
+    // The Tracer can be activated later
+    d_activated = false;
+#endif
+  }
+
+  /**
+   * @brief Add an attribute to the Trace
+   *
+   * @param key
+   * @param value
+   */
+  void setTraceAttribute(const std::string& key, const AnyValue& value);
+
+  /**
+   * @brief Create a new Span
+   *
+   * The Span's start time is set to the current time
+   *
+   * @param name The name for this span
+   * @return The SpanID of the created Span
+   */
+  SpanID addSpan(const std::string& name);
+
+  /**
+   * @brief Create a new Span with a parent
+   *
+   * The Span's start time is set to the current time
+   *
+   * @param name The name for this span
+   * @param parentSpanID The SpanID of the parent Span (not verified)
+   * @return The SpanID of the created Span
+   */
+  SpanID addSpan(const std::string& name, const SpanID& parentSpanID);
+
+  /**
+   * @brief Set an attribure on a Span
+   *
+   * This does not work when the Tracer is not active
+   *
+   * @param spanID The SpanID of the Span to add the attibute to
+   * @param key
+   * @param value
+   */
+  void setSpanAttribute(const SpanID& spanID, const std::string& key, const AnyValue& value);
+
+  /**
+   * @brief Sets the stop timestamp for a span
+   *
+   * When a Span is already closed, the timestamp is not updated
+   *
+   * @param spanID The ID of the Span to set the end time for
+   */
+  void closeSpan(const SpanID& spanID);
```

Comment:
I wonder if this should be private as well.

Reply:
`closeSpan` can't be private, `Closer` uses it.